Add ability to download resulting config

epodivilov opened the pull request:
Hi.

I and others (according to requests https://github.com/jakearchibald/svgomg/issues/11, https://github.com/jakearchibald/svgomg/issues/50 and https://github.com/jakearchibald/svgomg/issues/184), miss the possibility to apply SVGO configuration to multiple files. As I understood it, batch processing support is not expected, that's why I suggest an intermediate solution - a possibility to download the resulting config and then apply it using the cli interface of SVGO.

Unfortunately, the latest versions of SVGO don't support sending the config as text, that's why you'll have to create the config file as JS.

```bash
# Example of usage
npx svgo --config="./path/to/svgo.config.js" -f ./input -o ./output
```

XhmikosR commented on src/js/page/ui/settings.js:
```diff
@@ -83,6 +89,47 @@ export default class Settings {
+    const floatPrecision = Number(settings.floatPrecision);
+    const plugins = [];
+
+    for (const [name, active] of Object.entries(settings.plugins)) {
```
Most of this block seems to be duplicated in svgo-worker. Doesn't seem good from a maintenance perspective.

epodivilov replied:
I decided to copy the code first. I can rework it a bit and take out the repeating code.

XhmikosR replied:
I think that would help, yeah. Because this kind of duplication can be bad in the future.

Maybe we should just have a function to return the config with plugins somewhere in a common place.

@jakearchibald any suggestions?

epodivilov replied:
Moved the code for getting plugins into a separate utility

XhmikosR commented on src/js/utils/download.js:
```diff
@@ -0,0 +1,15 @@
+export function downloadSvgoConfig(filename, text) {
+  const element = document.createElement('a');
```
I'm wondering, isn't there a cleaner way without adding the extra `a` tag?

epodivilov replied:
Rewrote the code so that the link to the configuration was always on the page and was updated when the configuration was changed

XhmikosR commented:
hey, there. Thanks for continuing your work on this.

Can you please make sure there are no lint issues and the the export actually works?

XhmikosR commented on src/js/utils/download.js:
```diff
@@ -0,0 +1,3 @@
+export function createFileURL(data, type) {
```
This doesn't need to be a new file AFACT since it;s only used once.

XhmikosR commented:
This doesn't seem to work, no lint errors, no console errors but the config is not downloaded, at lean of Firefox that I tried.

epodivilov commented:
> This doesn't seem to work, no lint errors, no console errors but the config is not downloaded, at lean of Firefox that I tried.

What version of Firefox do you have? 

I have tested with the following versions on my system (MacOS Ventura) 
- Chrome 107.0.5304.87
- Firefox 106.0.4 
- Safari 16.1

I tested the export button on localhost and on the netlify build

XhmikosR commented:
Maybe you are missing a `click` event?

epodivilov commented:
> Maybe you are missing a `click` event?

But it's a link, it doesn't need click event handler. 
Maybe you have an old variant with button cached?

XhmikosR commented:
Yup, that was it.

Now, the only issue I see is ripple not firing when clicking the button.

XhmikosR commented on src/js/utils/settings.js:
```diff
@@ -0,0 +1,28 @@
+export function collectPlugins(settings) {
```
Just wondering, maybe `getActivePlugins` would be a better match?

XhmikosR commented:
This looks pretty good now! Waiting for @jakearchibald to review too.

KnifeFed commented:
No progress on the review? I guess https://deploy-preview-383--svgomg.netlify.app works for now though 🙂

AdrianoCahete commented:
JFYI: Using the preview build, I was able to export and use the config, but the property `cleanupIDs` returns an error:

```
Error: Unknown builtin plugin "cleanupIDs" specified.
    at resolvePluginConfig (...)
```

Mostly because `cleanupIDs` was renamed to `cleanupIds` in v3, as referenced in [changelog](https://github.com/svg/svgo/releases/tag/v3.0.0).

Manually renaming on config file works.

epodivilov commented:
> cleanupIds

Thanks for the feedback. Checked, there is indeed a problem. It is true that for SVGOMG to support the new plugin name you need to update the svgo package.

I updated my PR and added the necessary edits
